src/drivers/amd/rembrandt/acp_dma.c

```diff
@@ -0,0 +1,662 @@
+	dma_cntl = (acp_dma_cntl_0_t) dma_chan_reg_read(channel, ACP_DMA_CNTL_0);
+	/* Do the HW stop of the DMA */
+	/* set DMAChRst bit to stop the transfer */
+	dma_cntl.bits.dmachrun      = 0;
+	dma_cntl.bits.dmachiocen    = 0;
+	dma_chan_reg_write(channel, ACP_DMA_CNTL_0, dma_cntl.u32all);
+	ch_sts = (acp_dma_ch_sts_t) dma_reg_read(channel->dma, ACP_DMA_CH_STS);
+	if (ch_sts.bits.dmachrunsts & dmach_mask) {
+		/* set the reset bit for this channel to stop the dma transfer */
+		dma_cntl.bits.dmachrst	 = 1;
+		dma_chan_reg_write(channel, ACP_DMA_CNTL_0, dma_cntl.u32all);
+	}
+
+	while (delay_cnt > 0) {
+		ch_sts = (acp_dma_ch_sts_t) dma_reg_read(channel->dma, ACP_DMA_CH_STS);
+		if (!(ch_sts.bits.dmachrunsts & dmach_mask)) {
+			/* clear the reset flag after successfully stopping the dma transfer
+			 * and break from the loop
+			 */
+			dma_cntl.bits.dmachrst = 0;
+			dma_chan_reg_write(channel, ACP_DMA_CNTL_0, dma_cntl.u32all);
+			break;
+		}
+		delay_cnt--;
+	}
+	return 0;
+}
+
+static int acp_dma_start(struct dma_chan_data *channel)
+{
+	acp_dma_cntl_0_t dma_cntl;
+	acp_dma_ch_sts_t dma_sts;
+	uint32_t chan_sts;
+	int ret = 0;
+	struct timer *timer = timer_get();
+	uint64_t deadline = platform_timer_get(timer) +
+				clock_ms_to_ticks(PLATFORM_DEFAULT_CLOCK, 1) * 500 / 1000;
+	if (channel->status != COMP_STATE_PREPARE &&
+	    channel->status != COMP_STATE_SUSPEND)
+		return -EINVAL;
+	channel->status = COMP_STATE_ACTIVE;
+	/* Clear DMAChRun before starting the DMA Ch */
+	dma_cntl =  (acp_dma_cntl_0_t) dma_chan_reg_read(channel, ACP_DMA_CNTL_0);
+	dma_cntl.bits.dmachrun = 0;
+	dma_cntl.bits.dmachiocen  = 0;
+	dma_chan_reg_write(channel, ACP_DMA_CNTL_0, dma_cntl.u32all);
+
+	dma_cntl =  (acp_dma_cntl_0_t) dma_chan_reg_read(channel, ACP_DMA_CNTL_0);
+	dma_cntl.bits.dmachrun	 = 1;
+	dma_cntl.bits.dmachiocen = 0;
+
+	/* set dmachrun bit to start the transfer */
+	dma_chan_reg_write(channel, ACP_DMA_CNTL_0, dma_cntl.u32all);
+
+	/* poll for the status bit
+	 * to finish the dma transfer
+	 * then initiate call back function
+	 */
+	dma_sts = (acp_dma_ch_sts_t)dma_reg_read(channel->dma, ACP_DMA_CH_STS);
+	chan_sts = dma_sts.u32all & (1<<channel->index);
+	while (chan_sts) {
+		if (deadline < platform_timer_get(timer)) {
+			/* safe check in case we've got preempted after read */
+			if (chan_sts)
+				return 0;
+			tr_err(&acpdma_tr, "acp-dma: timed out for dma start");
+			return -ETIME;
+		}
+		dma_sts = (acp_dma_ch_sts_t)dma_reg_read(channel->dma, ACP_DMA_CH_STS);
+		chan_sts = dma_sts.u32all & (1<<channel->index);
+	}
+	return ret;
+}
+
+static int acp_dma_release(struct dma_chan_data *channel)
+{
+	tr_info(&acpdma_tr, "DMA: release(%d)", channel->index);
+	if (channel->status != COMP_STATE_PAUSED)
+		return -EINVAL;
+	channel->status = COMP_STATE_ACTIVE;
+	return 0;
+}
+
+static int acp_dma_pause(struct dma_chan_data *channel)
+{
+	tr_info(&acpdma_tr, "h/w pause is not supported, changing the status of(%d) channel",
```

Comment:
We don't have any DMA hw pause functionality. So callback is to update the state machine of channel current channel context. During Pause call back if DMA HW is running, then once transfer is complete it will move to stop.